File: ts/earthbucks-lib/src/pub-key.ts
import { Buffer } from "buffer";
import { IsoHex } from "./iso-hex.ts";
import bs58 from "bs58";
import { PrivKey } from "./priv-key.ts";
import * as Hash from "./hash.ts";
import { Result, Ok, Err } from "earthbucks-opt-res";
import {
  EbxError,
  InvalidChecksumError,
  InvalidEncodingError,
  NotEnoughDataError,
  TooMuchDataError,
} from "./ebx-error.ts";
import { Option, None, Some } from "earthbucks-opt-res";

export class PubKey {
  static readonly SIZE = 33; // y-is-odd byte plus 32-byte x

  buf: Buffer;

  constructor(buf: Buffer) {
    this.buf = buf;
  }

  static fromPrivKey(privKey: PrivKey): Result<PubKey, EbxError> {
    const res = privKey.toPubKeyBuffer();
    if (res.err) {
      return Err(res.val);
    }
    return Ok(new PubKey(res.unwrap()));
  }

  static fromIsoBuf(buf: Buffer): Result<PubKey, EbxError> {
    if (buf.length > PubKey.SIZE) {
      return Err(new TooMuchDataError(None));
    }
    if (buf.length < PubKey.SIZE) {
      return Err(new NotEnoughDataError(None));
    }
    return Ok(new PubKey(buf));
  }

  toIsoBuf(): Buffer {
    return this.buf;
  }

  toIsoHex(): string {
    return this.buf.toString("hex");
  }

  static fromIsoHex(hex: string): Result<PubKey, EbxError> {
    const res = IsoHex.decode(hex);
    if (res.err) {
      return Err(res.val);
    }
    const buf = res.unwrap();
    return PubKey.fromIsoBuf(buf);
  }

  toIsoStr(): string {
    const checkHash = Hash.blake3Hash(this.buf);
    const checkSum = checkHash.subarray(0, 4);
    const checkHex = checkSum.toString("hex");
    return "ebxpub" + checkHex + bs58.encode(this.buf);
  }

  static fromIsoStr(str: string): Result<PubKey, EbxError> {
    if (!str.startsWith("ebxpub")) {
      return Err(new InvalidEncodingError(None));
    }
    const checkHex = str.slice(6, 14);
    const res = IsoHex.decode(checkHex);
    if (res.err) {
      return Err(res.val);
    }
    const checkBuf = res.unwrap();
    let decoded: Buffer;
    try {
      decoded = Buffer.from(bs58.decode(str.slice(14)));
    } catch (e) {
      return Err(new InvalidChecksumError(None));
    }
    const checkHash = Hash.blake3Hash(decoded);
    const checkSum = checkHash.subarray(0, 4);
    if (!checkBuf.equals(checkSum)) {
      return Err(new InvalidEncodingError(None));
    }
    return PubKey.fromIsoBuf(decoded);
  }

  static isValidStringFmt(str: string): boolean {
    const res = PubKey.fromIsoStr(str);
    return res.ok;
  }
}

File: rs/earthbucks_lib/src/pub_key.rs
use crate::ebx_error::EbxError;
use crate::hash::blake3_hash;
use crate::iso_hex;
use crate::priv_key::PrivKey;
use bs58;
use secp256k1::PublicKey;

#[derive(Debug, Clone)]
pub struct PubKey {
    pub buf: [u8; PubKey::SIZE],
}

impl PubKey {
    pub const SIZE: usize = 33;

    pub fn new(pub_key: [u8; PubKey::SIZE]) -> Self {
        PubKey { buf: pub_key }
    }

    pub fn from_iso_buf(vec: Vec<u8>) -> Result<Self, EbxError> {
        if vec.len() > PubKey::SIZE {
            return Err(EbxError::TooMuchDataError { source: None });
        }
        if vec.len() < PubKey::SIZE {
            return Err(EbxError::NotEnoughDataError { source: None });
        }
        let mut pub_key = [0u8; PubKey::SIZE];
        pub_key.copy_from_slice(&vec);
        Ok(PubKey::new(pub_key))
    }

    pub fn to_buffer(&self) -> &[u8; PubKey::SIZE] {
        &self.buf
    }

    pub fn from_priv_key(priv_key: &PrivKey) -> Result<Self, EbxError> {
        let pub_key_buf = priv_key.to_pub_key_buffer();
        if pub_key_buf.is_err() {
            return Err(EbxError::InvalidKeyError { source: None });
        }
        Ok(PubKey::new(pub_key_buf.unwrap()))
    }

    pub fn to_iso_hex(&self) -> String {
        hex::encode(self.buf)
    }

    pub fn from_iso_hex(hex: &str) -> Result<PubKey, EbxError> {
        let pub_key_buf = iso_hex::decode(hex)?;
        PubKey::from_iso_buf(pub_key_buf)
    }

    pub fn to_iso_str(&self) -> String {
        let check_hash = blake3_hash(&self.buf);
        let check_sum = &check_hash[0..4];
        let check_hex = hex::encode(check_sum);
        "ebxpub".to_string() + &check_hex + &bs58::encode(&self.buf).into_string()
    }

    pub fn from_iso_str(s: &str) -> Result<PubKey, EbxError> {
        if !s.starts_with("ebxpub") {
            return Err(EbxError::InvalidEncodingError { source: None });
        }
        let check_str = &s[6..14];
        let check_buf = iso_hex::decode(check_str)?;
        let buf = bs58::decode(&s[14..])
            .into_vec()
            .map_err(|_| EbxError::InvalidEncodingError { source: None })?;
        let check_hash = blake3_hash(&buf);
        let check_sum = &check_hash[0..4];
        if check_buf != check_sum {
            return Err(EbxError::InvalidChecksumError { source: None });
        }
        PubKey::from_iso_buf(buf)
    }

    pub fn is_valid(&self) -> bool {
        let public_key_obj = PublicKey::from_slice(&self.buf);
        public_key_obj.is_ok()
    }

    pub fn is_valid_string_fmt(s: &str) -> bool {
        let res = Self::from_iso_str(s);
        res.is_ok()
    }
}
